static/src/javascripts/projects/commercial/modules/messenger.ts

```diff
@@ -0,0 +1,409 @@
+ * on occasion be larger than the size returned by DFP. And so they
+ * have been setup to send a message of the form:
+ */
+type ProgrammaticMessage = {
+	type: string;
+	value: {
+		id?: string;
+		slotId?: string;
+		height: number;
+		width: number;
+	};
+};
+
+/**
+ * Callbacks that can be registered to fire when receiving messages from an iframe
+ */
+type ListenerCallback = (
+	/**
+	 * The data payload sent by an iframe, and has type `unknown` because we can't
+	 * predict what the iframe will send. It is the responsibility of the callback
+	 * to obtain a value of the desired type (or fail gracefully)
+	 */
+	specs: unknown | undefined,
+	/**
+	 * Non-persistent callbacks can be chained together. This value is the return
+	 * value of the previously fired callback in the chain. It is the responsibility
+	 * of the current callback to either ignore it or use it / pass along
+	 */
+	ret: unknown,
+	/**
+	 * Reference to the iframe that is the source of the message
+	 */
+	iframe?: HTMLIFrameElement,
+) => unknown;
+
+/**
+ * The set of listeners currently registered
+ *
+ * Each message sent by an iframe has a `type` field which indicates the kind of
+ * message e.g. `resize`, `measure-ad-load`. One or more listeners is registered
+ * for each type.
+ */
+type Listeners = Partial<
+	Record<MessageType, ListenerCallback | ListenerCallback[] | undefined>
+>;
+
+/**
+ * Types of functions to register a listener for a given type of iframe message
+ */
+export type RegisterListener = (
+	type: MessageType,
+	callback: ListenerCallback,
+	options?: {
+		window?: WindowProxy;
+		persist?: boolean;
+	},
+) => void;
+
+/**
+ * Types of functions to unregister a listener for a given type of iframe message
+ *
+ */
+export type UnregisterListener = (
+	type: MessageType,
+	callback?: ListenerCallback,
+	options?: {
+		window?: WindowProxy;
+	},
+) => void;
+
+const LISTENERS: Listeners = {};
+let REGISTERED_LISTENERS = 0;
+
+const error405 = {
+	code: 405,
+	message: 'Service %% not implemented',
+};
+const error500 = {
+	code: 500,
+	message: 'Internal server error\n\n%%',
+};
+
+/**
+ * Determine if an unknown payload has the shape of a programmatic message
+ *
+ * @param payload The unknown message payload
+ */
+const isProgrammaticMessage = (
+	payload: unknown,
+): payload is ProgrammaticMessage => {
+	const payloadToCheck = payload as ProgrammaticMessage;
+	return (
+		payloadToCheck.type === 'set-ad-height' &&
+		('id' in payloadToCheck.value || 'slotId' in payloadToCheck.value) &&
+		'height' in payloadToCheck.value
+	);
+};
+
+/**
+ * Convert a legacy programmatic message to a standard message
+ *
+ * Note that this only applies to specific resize programmatic messages
+ * (these include specific width and height values)
+ */
+const toStandardMessage = (
+	payload: ProgrammaticMessage,
+): StandardMessage<{ width: number; height: number }> => ({
+	id: 'aaaa0000-bb11-cc22-dd33-eeeeee444444',
```

Comment:
Not blocking: I wonder if there's a cleaner way to skip the validation step for this property of legacy programmatic messages without adding the magic uuid.
Maybe worth adding a comment with a TODO or at least explaining what's going on here?

Reply:
By validation are you referring to the regex test in `isValidPayload`? Because that only gets called if the message wasn't programmatic to begin with, so this specific UUID won't get validated.

Reply:
Yes, I meant `isValidPayload` - I'd missed that we were no longer calling isValidPayload for programmatic messages! I left a separate comment about that.

Reply:
For posterity–this is where the UUID is generated: https://github.com/guardian/commercial-templates/blob/svelte-kit/legacy/src/_shared/js/messages.js#L180